lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+	}
+
+	// Here's where it gets a little tricky. If the parent has a status of ValidateFailed, then we know we store
+	// this block as ValidateFailed. If the parent is not ValidateFailed, we ONLY store the block and move on.
+	// We don't want to store it as ValidateFailed because we don't know if it's actually invalid.
+	if parentBlockNode.IsValidateFailed() {
+		return bc.storeValidateFailedBlockWithWrappedError(block, errors.New("parent block is ValidateFailed"))
+	}
+
+	// If the parent block still has a Stored status, it means that we weren't able to validate it
+	// despite trying. The current block will also be stored as a Stored block.
+	if !parentBlockNode.IsValidated() {
+		return bc.storeValidateFailedBlockWithWrappedError(block, errors.New("parent block is neither Validated nor ValidateFailed"))
+	}
+
+	// At this point, we know the parent block is validated. We can now perform all other validations
+	// on the current block. First we store the block in the block index.
+	blockNode, err = bc.storeBlockInBlockIndex(block)
+	if err != nil {
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem adding block to block index: %v", err)
+	}
+
+	// Check if the block is properly formed and passes all basic validations.
+	if err = bc.isValidBlockPoS(block); err != nil {
+		return bc.storeValidateFailedBlockWithWrappedError(block, err)
+	}
+
+	// We expect the utxoView for the parent block to be valid because we check that all ancestor blocks have
+	// been validated.
+	utxoView, err := bc.getUtxoViewAtBlockHash(*block.Header.PrevBlockHash)
+	if err != nil {
+		// This should never happen. If the parent is validated and extends from the tip, then we should
+		// be able to build a UtxoView for it. This failure can only happen due to transient or badger issues.
+		// We return the block node as-is as a best effort thing.
+		return blockNode, nil
+	}
+	// A couple of options on how to handle:
+	//   - Add utility to UtxoView to fetch the validator set given an arbitrary block height. If we can't fetch the
+	//     validator set for the block, then we reject it (even if it later turns out to be a valid block)
+	//   - Add block to the block index before QC validation such that even if we aren't able to fetch the validator
+	//     set for the block, we can at least store it locally.
+	// 2. Validate QC
+	validatorsByStake, err := utxoView.GetAllSnapshotValidatorSetEntriesByStake()
+	if err != nil {
+		// This should never happen. If the parent is validated and extends from the tip, then we should
+		// be able to fetch the validator set at its block height for it. This failure can only happen due
+		// to transient badger issues. We return the block node as-is as a best effort thing.
+		return blockNode, nil
+	}
+
+	// Validate the block's QC. If it's invalid, we store it as ValidateFailed.
+	if err = bc.isValidPoSQuorumCertificate(block, validatorsByStake); err != nil {
```

Comment:
I'm skipping over this function. Lmk if you want me to look at it very carefully.

Reply:
Yes please look at it. This is the 2/3s supermajority QC check which is pretty critical.